### PR TITLE
[Qwen3-Omni Perf] Talker AR loop: sampling-state reuse

### PR DESCRIPTION
**Pull request by AkazaAkane**
<html><head></head><body><h2>Motivation</h2><p><code>Qwen3OmniTalker.prepare_decode_buffers()</code> previously rebuilt sampling tensors and repetition/suppression masks on every autoregressive decode step, even when the active request batch and sampling parameters were unchanged.</p><p>This introduced repeated Python-side list construction, mask reconstruction, tensor allocation, and host-to-device transfers in the Talker AR loop.</p><p>This PR adds a steady-state fast path that reuses the existing decode buffers when the request batch remains stable. It also consolidates sampling-parameter transfers through reusable staging buffers.</p><h2>Modifications</h2><ul><li><p>Add a steady-state reuse path for <code>prepare_decode_buffers()</code>.</p></li><li><p>Cache:</p><ul><li><p>request IDs;</p></li><li><p>output lengths;</p></li><li><p>rows with active repetition penalties.</p></li></ul></li><li><p>Reuse the existing sampling parameters and suppression mask when:</p><ul><li><p>batch size is unchanged;</p></li><li><p>request ordering and IDs are unchanged;</p></li><li><p>each request has advanced by exactly one output token.</p></li></ul></li><li><p>Incrementally update only the repetition-mask entries corresponding to newly sampled tokens.</p></li><li><p>Fall back to a full buffer rebuild when the cached batch state is no longer valid.</p></li><li><p>Add reusable pinned CPU and GPU staging buffers for sampling parameters.</p></li><li><p>Consolidate repetition penalty, temperature, <code>top_p</code>, <code>top_k</code>, <code>min_p</code>, and sampling seed into one staged host-to-device transfer.</p></li><li><p>Add <code>invalidate_decode_buffers()</code> so prefill/extend operations that bypass the normal decode update path cannot incorrectly enter the reuse path.</p></li><li><p>Add unit tests covering:</p><ul><li><p>steady-state reuse;</p></li><li><p>repetition-mask updates;</p></li><li><p>request-ID and batch changes;</p></li><li><p>output-length discontinuities;</p></li><li><p>explicit invalidation after prefill;</p></li><li><p>sampling-state rebuild behavior.</p></li></ul></li></ul><h2>Related Issues</h2><p>Refs #1025.</p><h2>Accuracy Test</h2><p>Correctness was validated in both isolated codec-stream comparisons and end-to-end SeedTTS runs.</p><h3>Codec-token comparison</h3><p>An earlier fixed-seed comparison covering <code>27 × 16</code> codec streams produced bit-identical codec tokens between baseline and this PR.</p><h3>End-to-end validation</h3><p>Across all six benchmark runs:</p><ul><li><p>every run completed <code>50/50</code> requests;</p></li><li><p>every run produced exactly <code>733</code> output tokens;</p></li><li><p>no output-token count differences were observed between baseline and this PR.</p></li></ul><p>Together, these results establish codec-token invariance for the tested workload.</p><h2>Benchmark &amp; Profiling</h2><p>Three interleaved, fixed-seed benchmark rounds were run for baseline and this PR.</p><h3>Test configuration</h3><pre><code class="language-text">Dataset: same 50 SeedTTS samples
Seed: 12345
Concurrency: 8 and 16
Thinker GPU: H100 GPU 2
Talker GPU: H100 GPU 3
Topology: BF16 disaggregated
Startup: deterministic heavy-stage startup
</code></pre><p>The disaggregated topology was used to eliminate colocated startup OOM variance. Round 3 reversed the baseline/PR execution order to reduce ordering bias.</p><h3>Results</h3>

Concurrency | Round | Baseline req/s | PR req/s | Delta
-- | -- | -- | -- | --
8 | 1 | 5.026 | 6.255 | +24.5%
8 | 2 | 5.994 | 6.232 | +4.0%
8 | 3, reverse order | 6.053 | 6.154 | +1.7%
16 | 1 | 8.377 | 9.093 | +8.6%
16 | 2 | 9.261 | 9.186 | −0.8%
16 | 3, reverse order | 9.120 | 9.322 | +2.2%

<h3>Interpretation</h3><ul><li><p>At concurrency 8, the PR provides a small repeatable throughput improvement of approximately <strong>2.8%</strong> after excluding the cold first pair.</p></li><li><p>At concurrency 16, throughput changes by approximately <strong>0.7%</strong>, which is effectively neutral within the observed run-to-run variance.</p></li><li><p>The optimization does not regress saturated throughput in the tested configuration.</p></li><li><p>The larger first-round improvements should not be treated as the headline result because they were not reproduced after warm startup.</p></li></ul><p>The defensible end-to-end result is therefore:</p><pre><code class="language-text">c8:  +2.8%
c16: +0.7%, effectively neutral
</code></pre>

## Detailed paired waveform comparison

Configuration:

| Item | Value |
| --- | --- |
| Baseline | 7dcb0ec0 (origin/main) |
| PR | 224ff15d |
| Hardware | NVIDIA H100 80 GB |
| Model | Qwen3-Omni-30B-A3B-Instruct |
| Topology | BF16 disaggregated |
| Dataset | SeedTTS-50 EN, voice cloning |
| Requests | 50 paired samples |
| Concurrency | 8 |
| Sampling seed | 12345 |
| Output format | 24 kHz, mono WAV |

Both arms ran concurrently on separate GPU pairs with identical request payloads.

| Waveform metric | Result |
| --- | --- |
| Successfully generated pairs | 50/50 |
| Identical WAV byte streams | 50/50 |
| Identical WAV SHA-256 hashes | 50/50 |
| Identical decoded PCM arrays | 50/50 |
| Identical sample rates | 50/50 |
| Identical channel counts | 50/50 |
| Identical frame counts | 50/50 |
| Maximum absolute PCM error | 0.0 |
| Maximum PCM RMSE | 0.0 |
| Minimum pair correlation | 1.0 |
| Pair SNR | ∞ |
| Total audio compared per arm | 170.09625 seconds |
| Total PCM frames per arm | 4,082,310 |
| Shortest clip | 18,645 frames / 0.776875 s |
| Longest clip | 154,365 frames / 6.431875 s |

Thus, the paired outputs are not merely similar: every saved baseline WAV is byte-for-byte identical to its corresponding PR WAV. Decoding the files and comparing samples independently also produces zero difference.

## Compared WER results

Both sets were independently submitted to the same Qwen/Qwen3-ASR-1.7B server.

| WER metric | Baseline | PR | Difference |
| --- | --- | --- | --- |
| Evaluated | 50/50 | 50/50 | 0 |
| Skipped | 0 | 0 | 0 |
| Corpus WER | 1.2411% | 1.2411% | 0 |
| Mean per-sample WER | 1.1302% | 1.1302% | 0 |
| Median per-sample WER | 0% | 0% | 0 |
| Per-sample WER std. dev. | 3.7829% | 3.7829% | 0 |
| Per-sample WER p95 | 8.50% | 8.50% | 0 |
| Maximum per-sample WER | 20.00% | 20.00% | 0 |
| Samples above 50% WER | 0 | 0 | 0 |
| Exact normalized transcript pairs | 50/50 | 50/50 | — |
| Exact per-sample WER pairs | 50/50 | 50/50 | — |

Corpus totals for both arms:

- Reference words: 564
- Correct words: 558
- Substitutions: 5
- Deletions: 1
- Insertions: 1
- Corpus WER: (5 + 1 + 1) / 564 = 1.2411%
- Forty-five samples had zero WER.

### Five non-zero-WER samples

| Sample | Reference → ASR hypothesis | Baseline | PR |
| --- | --- | --- | --- |
| common_voice_en_123125-common_voice_en_123126 | “who doesn t” → “who does not” | 14.2857% | 14.2857% |
| common_voice_en_155313-common_voice_en_155315 | “tim tebow” → “tim tibo” | 5.5556% | 5.5556% |
| common_voice_en_20048255-common_voice_en_20048259 | “queen” → “queens” | 10.0000% | 10.0000% |
| common_voice_en_37232389-common_voice_en_37190783 | “will you…” → “we will…” | 20.0000% | 20.0000% |
| common_voice_en_597578-common_voice_en_597583 | “wilma s” → “wilma has” | 6.6667% | 6.6667% |

The normalized hypotheses, error counts, and WER values are identical between baseline and PR for every sample.

### Codec-token identity test

| Metric | Baseline | PR #1043 | Comparison |
| --- | --- | --- | --- |
| Paired requests | 50 | 50 | 50/50 paired |
| Successful requests | 50 | 50 | 100% |
| Talker frames | 2,254 | 2,254 | Equal |
| Codebooks per frame | 16 | 16 | Equal |
| Codec-token IDs | 36,064 | 36,064 | Equal |
| Exact token IDs | — | — | 36,064/36,064 |
| Mismatched token IDs | — | — | 0 |
| Exact request tensors | — | — | 50/50 |
| Equal tensor shapes | — | — | 50/50 |
| Equal EOS positions | — | — | 50/50 |
| Equal per-request SHA-256 | — | — | 50/50 |
| First token mismatch | — | — | None |
| Completion-token count | 733 | 733 | Equal |
| Byte-identical WAVs | — | — | 50/50 |
| PCM-identical waveforms | — | — | 50/50 |
| Maximum absolute PCM error | — | — | 0.0 |
| Maximum PCM RMSE | — | — | 0.0 |
| Minimum waveform correlation | — | — | 1.0 |

### WER comparison

| Metric | Baseline | PR #1043 | Difference |
| --- | --- | --- | --- |
| Evaluated samples | 50/50 | 50/50 | 0 |
| Skipped samples | 0 | 0 | 0 |
| Corpus WER | 1.2411% | 1.2411% | 0 |
| Mean per-sample WER | 1.1302% | 1.1302% | 0 |
| Median per-sample WER | 0.0000% | 0.0000% | 0 |
| Per-sample WER p95 | 8.5000% | 8.5000% | 0 |
| Maximum per-sample WER | 20.0000% | 20.0000% | 0 |
| Samples above 50% WER | 0 | 0 | 0 |
| Exact normalized transcripts | — | — | 50/50 |
| Exact per-sample WER values | — | — | 50/50 |

### Test configuration

| Parameter | Value |
| --- | --- |
| Baseline commit | 7dcb0ec0 |
| PR commit | 224ff15d |
| Model | Qwen3-Omni-30B-A3B-Instruct |
| Dataset | SeedTTS-50 English |
| Hardware | NVIDIA H100 80 GB |
| Precision/topology | BF16 disaggregated |
| Sampling seed | 12345 |
| Token-capture concurrency | 1, controlled batch composition |
| Waveform/WER concurrency | 8 |
| ASR model | Qwen/Qwen3-ASR-1.7B |
| Token capture point | Talker output immediately before code2wav |

• Full pooled streaming A/B results:

| Concurrency | Metric | Baseline | PR | Delta |
| :--- | :--- | :--- | :--- | :--- |
| 8 | Requests | 150 | 150 | — |
| 8 | Request throughput (req/s) | 5.426 | 5.905 | +8.8% |
| 8 | TTFA p50 (s) | 0.557 | 0.541 | −2.8% |
| 8 | TTFA p95 (s) | 1.693 | 1.341 | −20.8% |
| 8 | Total latency p50 (s) | 1.327 | 1.262 | −4.9% |
| 8 | Total latency p95 (s) | 2.273 | 2.175 | −4.3% |
| 8 | xRT (decoded audio s / wall s) | 18.791 | 20.453 | +8.8% |
| 8 | RTF p50 | 0.381 | 0.360 | −5.5% |
| 8 | RTF p95 | 0.815 | 0.758 | −7.0% |
| 8 | Chunks/request, mean | 4.80 | 4.80 | 0 |
| 8 | Decoded duration/request, mean (s) | 3.4634 | 3.4634 | 0 |
| 8 | Total decoded duration (s) | 519.509 | 519.509 | 0 |
| 16 | Requests | 250 | 250 | — |
| 16 | Request throughput (req/s) | 5.771 | 5.834 | +1.1% |
| 16 | TTFA p50 (s) | 1.564 | 1.694 | +8.3% |
| 16 | TTFA p95 (s) | 2.703 | 2.719 | +0.6% |
| 16 | Total latency p50 (s) | 2.560 | 2.535 | −1.0% |
| 16 | Total latency p95 (s) | 4.030 | 3.739 | −7.2% |
| 16 | xRT (decoded audio s / wall s) | 19.994 | 20.208 | +1.1% |
| 16 | RTF p50 | 0.735 | 0.746 | +1.4% |
| 16 | RTF p95 | 1.270 | 1.202 | −5.3% |
| 16 | Chunks/request, mean | 4.80 | 4.80 | 0 |
| 16 | Decoded duration/request, mean (s) | 3.4644 | 3.4640 | −0.01% |
| 16 | Total decoded duration (s) | 866.089 | 866.009 | −0.01% |

Paired request-throughput results:

| Concurrency | Round | Baseline req/s | PR req/s | Delta |
| :--- | :--- | :--- | :--- | :--- |
| 8 | 1 warm | 5.122 | 5.753 | +12.3% |
| 8 | 2 warm | 5.649 | 6.093 | +7.9% |
| 8 | 3 warm | 5.535 | 5.880 | +6.2% |
| 16 | 1 | 5.913 | 5.908 | −0.1% |
| 16 | 2 | 5.481 | 5.761 | +5.1% |
| 16 | 3 | 6.368 | 5.561 | −12.7% |
| 16 | 4 | 5.645 | 6.387 | +13.1% |
| 16 | 5 | 5.537 | 5.622 | +1.5% |
| 8 | Median delta | — | — | +7.9% |
| 16 | Median delta | — | — | +1.5% |

TTFA is time to the first decoded audio chunk. RTF is request latency divided by decoded duration; lower is better.
xRT is total decoded duration divided by benchmark wall time; higher is better.

## Checklist

* [x] Format your code according with pre-commit.
* [x] Add unit tests.
* [x]  Update documentation / docstrings / example tutorials as needed.
* [x]  Provide throughput / latency benchmark results and accuracy evaluation results as needed.
* [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

## CI

CI runs on self-hosted GPU runners and requires a maintainer to add the `run-ci` label. Once labeled, every subsequent push re-triggers CI as long as the label remains. Use `/tag-and-rerun-ci higgs` or `/tag-and-rerun-ci moss` to select a TTS CI model. Draft PRs are skipped even if labeled.

### CODE REVIEW

**Comment by edwingao28** [comment]
Thanks — nice work. We ran an independent attribution pass on H200 (colocated pipeline, c8, torch profiler on both arms, identical bench, base = recent main + this PR merged) to see exactly what the reuse removes per talker AR frame:

| per talker AR frame (c8 steady) | baseline | this PR | delta |
|---|---|---|---|
| tiny pageable HtoD uploads | 15.1 | 4.3 | −72% |
| forward-thread cudaStreamSynchronize | 16.1 | 5.2 | −67% |
| total cudaMemcpyAsync | 44.8 | 26.3 | −41% |

Controls were clean: MoE GEMM count identical, DtoH unchanged, CUDA graph replay share unchanged, codec chunk structure unchanged. So the PR removes the whole sampling-state rebuild cluster, exactly as advertised. Caveat: these are profiler-on attribution counts, not wall-clock claims — the E2E side is covered by the fixed-seed numbers in the PR description.

Follow-ups we measured that this PR deliberately doesn't touch (none block this PR; will fold into #1022):

- [ ] Batch the 16 per-request output `.clone()`s in chunk emission into 2 batched clones (~14 memcpys/frame)
- [ ] Drop the second `.tolist()` on the same token ids (the omni output processor already syncs them)
- [ ] Steady-state fast path for `_write_feedback_buffers` (cached device arange removes one pageable HtoD + sync per frame)
- [ ] Stage the token readback through a pinned buffer instead of a pageable `.tolist()`

File: tests/unit_test/qwen3_omni/test_talker.py [hunk 1]
```diff
@@ -2166,7 +2166,19 @@ def _talker_seed_self(max_bs: int = 4, vocab: int = 8) -> SimpleNamespace:
         _sampling_top_ks=torch.ones(max_bs, dtype=torch.long),
         _sampling_min_ps=torch.zeros(max_bs),
         _sampling_seeds=torch.zeros(max_bs, dtype=torch.long),
+        _sampling_staging_cpu=torch.zeros(6, max_bs, dtype=torch.int64),
+        _sampling_staging_gpu=torch.zeros(6, max_bs, dtype=torch.int64),
+        _sampling_staging_event=None,
```
Comment (edwingao28) [inline tests/unit_test/qwen3_omni/test_talker.py]: The new behavior is CUDA-specific but this fixture uses CPU tensors and sets `_sampling_staging_event=None`, so the added tests never execute that path. 

Please add a CUDA regression test that runs consecutive rebuild/reuse cycles, synchronizes the stream, and compares every sampling tensor and mask with a fresh preparation.

File: tests/unit_test/qwen3_omni/test_talker.py [hunk 2]
```diff
@@ -2212,3 +2224,97 @@ def test_talker_prepare_decode_buffers_unseeded_seed_is_rank_shared() -> None:
...
+
+    fake, requests = _prepared()
+    _advance(requests)
+    requests[0].data.req.output_ids.append(6)
+    Qwen3OmniTalker.prepare_decode_buffers(fake, requests)
+    assert float(fake._sampling_temperatures[0, 0]) == pytest.approx(0.8)
+
+    fake, requests = _prepared()
+    _advance(requests)
+    fake.invalidate_decode_buffers()
```
Comment (edwingao28) [inline tests/unit_test/qwen3_omni/test_talker.py]: This test calls invalidate_decode_buffers() directly, so it only tests the helper not whether the prefill path actually triggers it. Please add a test that runs an extend/prefill step followed by decode， that would catch regressions where prefill stops invalidating the cached sampling state.

**Review by edwingao28 — COMMENTED** [review]
@AkazaAkane Left some comments. The PR body reports an earlier token comparison and equal output-token counts, but equal counts do not establish token identity, and no waveform or WER result is included. Please attach the paired waveform comparison and compred WER results.

File: tests/unit_test/qwen3_omni/test_talker.py (hunk 1 above)
Comment (AkazaAkane) [reply tests/unit_test/qwen3_omni/test_talker.py]: Added device-aware Talker fixtures and a CUDA rebuild/reuse equivalence test in test_talker.py (line 2166).

File: tests/unit_test/qwen3_omni/test_talker.py (hunk 2 above)
Comment (AkazaAkane) [reply tests/unit_test/qwen3_omni/test_talker.py]: Added an extend/prefill → decode integration test verifying cache invalidation.

**Comment by AkazaAkane** [comment]
FYI @edwingao28 All test result are updated. Comparison result suggest quality unchanged.

**Review by SandyLuXY — COMMENTED** [review]
Nice PR! Non-blocking comment: the current A/B benchmarking supports request throughput only, consider including TTFA and total-latency p50/p95, xRT, RTF p50/p95, chunks/request, decoded duration etc in the PR body.

**Comment by AkazaAkane** [comment]
> Nice PR! Non-blocking comment: the current A/B benchmarking supports request throughput only, consider including TTFA and total-latency p50/p95, xRT, RTF p50/p95, chunks/request, decoded duration etc in the PR body.

Added to PR body.

**Review by edwingao28 — APPROVED** [review]

**Comment by edwingao28** [comment]
lgtm cc: @SandyLuXY @yxs
